### ISSUE TITLE
Chalice Deploy Rewrite

### ISSUE THREAD

**Issue by joguSD** [aws/chalice#604]
## Deployer Rewrite

This is a tracking issue for reworking the deployment system in chalice.
To summarize the key points of this work:

* Generalize the deployments to not be tightly coupled to API gateway and
  Lambda resources.  This will allow for features such as Lambda only chalice apps
  without any API Gateway components.
* Allow chalice to support more AWS resources
  (https://github.com/aws/chalice/issues/516)
* Allow for smarter/smaller update "diffs"
* Allow for deployment parallelism (being able to deploy resources in parallel)

This will also have the benefit of being easier to troubleshoot
and maintain for chalice devs because all the deployment phases are extracted
out into separate classes.

This work should not result in any backwards incompatible changes, except in a
few cases now that are either undocumented or buggy.

This work is being tracked in the
[new-deployer](https://github.com/aws/chalice/tree/new-deployer) branch.

Related issues:

* #516 - Reduces the effort needed to implement this proposal
* #570 - Chalice without lambdas
* #162 - Lambda without API Gateway 
* #640 - Lambda without API Gateway (dupe)
* #488 - SNS triggers.  Updated deployments will make new event sources easier to implement

**Comment by nplutt** [comment]
Hey not sure if this is the place for this, but I think that it would be very useful if you could deploy a lambda function without the requirement of having an API gateway be deployed with it as well.

Thanks,
Nick

**Comment by bubba-h57** [comment]
My scheduled tasks need a different **lambda_timeout** & **lambda_memory_size** than the application/stage default.

It would be nice to see the same naming convention used in the API Gateway as that which is used in Lambda.

It would be nice to override the default naming convention(s) in the **config.json**.

**Comment by wollerman** [comment]
This references #488 but doesn't seem to address it. Is that task still part of this?

**Comment by jamesls** [comment]
That's still a planned feature I'd like to implement (and am currently working on) but it's not part of this specific issue.
